src/backend/buffer.rs

```diff
@@ -200,6 +207,16 @@ where
     }
 }
 
+/// Calculate var7 encoded size from a given value.
+#[inline]
+fn calculate_var7_size(value: usize) -> usize {
+    // number of bits to encode
+    // value = 0 would give 0 bits, hence: |1, could be anything up to |0x7F as well
+    let bits = usize::BITS - (value | 1).leading_zeros();
+    // (bits to encode / 7).ceil()
+    ((bits + 6) / 7) as usize
+}
```

Comment:
Can you please explain this formula a bit better so that it does not look like magic to the reader?

Reply:
Also: is this precise or an over approximation?

Reply:
This formula is precise and this is also reflected in tests.

Reply:
The formula goes like this:

1. Take the binary size (bit length) of the original length

```
let bits = usize::BITS - (value | 1).leading_zeros();
```
For example if your length is `42` this is binary `101010`, the bit length is 6.
Assuming `usize::BITS` is 64 `leading_zeros` of 42usize is 58. 64 - 58 = 6.

2. If the binary size is 0, make it 1 (0 is the only case that breakes the formula, hence `(value | 1)`).

3. The encoded size in bytes is the total number of encoded bits divided by 7 (each byte encodes only 7 bits of the original value), rounded up because the last byte's 7-bits might not be filled entirely: hence (bits + 6).

```
// (bits to encode / 7).ceil()
((bits + 6) / 7) as usize
```

For example 0x7F bit length is 7, (7 + 6) / 7 == 1

The length is encoded as: 0b0111_1111

0x80 bit length is 8, (8 + 6) / 7 == 2

The length is encoded as: 0b1000_0000 0b0000_0001

Reply:
Let's take the original bit value of the encoded length (each bit placement is represented by a character)

ZYXWVUTSRQPONMLKJIHGFEDCBA9876543210 (bit-length: 36)

The encoded chunks end up being:

!6543210 !DCBA987 !KJIHGFE !RQPONML !YXWVUTS _______Z

Where `!` and `_` are `1` and `0` bit markers respectively.

As seen clearly each 7 bits ends up in each chunk with the last chunk filled with the remaining 1-7 bits

(36 + 6) / 7 == 6

Now if we take:

YXWVUTSRQPONMLKJIHGFEDCBA9876543210 (bit-length: 35)

The encoded chunks end up being:

!6543210 !DCBA987 !KJIHGFE !RQPONML _YXWVUTS

(35 + 6) / 7 == 5

This process repeats every 7 bits.

The only exception is the `0` itself which if we followed the formula precisely should be encoded in exactly `0` bytes - because there are no bits to encode!

This leads us to the interesting observation that we could in fact extend the encoded var7 by adding any number of redundant `0` at the end, and the decoding formula (providing we have an infinite integer capacity) would still work:

!6543210 !DCBA987 !KJIHGFE !RQPONML !YXWVUTS ________
!6543210 !DCBA987 !KJIHGFE !RQPONML !YXWVUTS !_______ ________
!6543210 !DCBA987 !KJIHGFE !RQPONML !YXWVUTS !_______ !_______ ________

and so on...

I hope that explains why `0` is breaking the formula without actually really breaking it. What we really need is just 1 bit of information that this is the end of the stream of data.